controllers/usersController.js

```diff
@@ -1,129 +1,219 @@
+		User.findById(userId)
+			.then((user) => {
+				// render the found users' data on edit page
+				res.render('users/edit', {
+					user: user,
+				});
+			})
+			.catch((error) => {
+				console.log(`Error fetching user by ID: ${error.message}`);
+				next(error);
+			});
+	},
+	update: (req, res, next) => {
+		let userId = req.params.id,
+			userParams = getUserParams(req.body);
+		// find a user and update the users' data
+		User.findByIdAndUpdate(userId, {
+			$set: userParams,
+		})
+			.then((user) => {
+				// console.log(`found user: ${user}`);
+				// add updated user data to a local variable
+				req.flash('updated', `${user.fullName}'s account updated successfully`);
+				res.locals.redirect = `/users/${userId}`;
+				res.locals.user = user;
+				// call middleware
+				next();
+			})
+			.catch((error) => {
+				console.log(`Error updating user by ID: ${error.message}`);
+				res.locals.redirect = `/users/${userId}/edit`;
+				req.flash(
+					'error',
+					`Failed to update user account because: ${error.message}.`,
+				);
+				next();
+			});
+	},
+	delete: (req, res, next) => {
+		let userId = req.params.id;
+		User.findByIdAndDelete(userId)
+			.then(() => {
+				req.flash('deleted', `${userId} has been deleted successfully`);
+				res.locals.redirect = '/users';
+				next();
+			})
+			.catch((error) => {
+				console.log(`Error deleting user by ID: ${error.message}`);
+				res.locals.redirect = '/users';
+				req.flash(
+					'error',
+					`Failed to delete user account because: ${error.message}`,
+				);
+				next();
+			});
+	},
+	login: (req, res) => {
+		res.render('users/login');
+	},
+	authenticate: (req, res, next) => {
+		User.findOne({
+			email: req.body.email,
+		})
+			.then((user) => {
+				if (user) {
+					user.passwordComparison(req.body.password).then((passwordMatch) => {
+						if (passwordMatch) {
+							res.locals.redirect = `/users/${user.id}`;
+							req.flash('success', `${user.fullName}'s logged in successfully`);
+							res.locals.user = user;
+						} else {
+							req.flash(
+								'error',
+								'Failed to log in user account: Incorrect Password.',
+							);
+							res.locals.redirect = '/users/login';
+						}
+						next();
+					});
+				} else {
+					req.flash(
+						'error',
+						'Failed to log in user account: User account not found.',
+					);
+					res.locals.redirect = '/users/login';
+					next();
+				}
+			})
+			.catch((error) => {
+				console.log(`Error logging in user: ${error.message}`);
+				next(error);
+			});
+	},
+	validate: [
+		sanitizeBody('email')
+			.normalizeEmail({
+				// change email characters to lowercase and trim unnecessary space
+				all_lowercase: true,
+			})
+			.trim(),
+		check('email', 'Email is invalid').isEmail(),
+		// check zipCode field
+		check('zipCode', 'Zip code is invalid').notEmpty().isInt().isLength({
+			min: 7,
+			max: 7,
+		}),
+		// check password field
+		check('password', 'Password cannot be empty').notEmpty(),
+	],
```

Comment:
Since express-validator v6, you should pass an array of validation for the validating middleware.
I created another middleware, getValidationResult() to set `req.skip = true;` and flash message.